### PR TITLE
Block type generator CLI command

### PR DESCRIPTION
**Pull request by jbrinley**
Generates all of the files necessary to stub out a new custom block type, in both the core plugin and the theme.

Usage:

```
wp s1 generate block <name>
```

E.g.,

```
wp s1 generate block image-gallery
```

### CODE REVIEW

**Comment by jbrinley** [comment]
An example of the dry-run output:

```
$ wp s1 generate block pineapple-pizza --dry-run
[Dry Run] Block config file: /application/www/wp-content/plugins/core/src/Blocks/Types/Pineapple_Pizza/Pineapple_Pizza.php
Block config contents:
<?php
declare( strict_types=1 );

namespace Tribe\Project\Blocks\Types\Pineapple_Pizza;

use Tribe\Libs\ACF\Block;
use Tribe\Libs\ACF\Block_Config;
use Tribe\Libs\ACF\Field;
use Tribe\Libs\ACF\Field_Section;

class Pineapple_Pizza extends Block_Config {
	public const NAME = 'pineapplepizza';

	public const SECTION_CONTENT  = 's-content';
	public const SECTION_SETTINGS = 's-settings';

	public const TITLE       = 'title';
	public const DESCRIPTION = 'description';

	public function add_block() {
		$this->set_block( new Block( self::NAME, [
			'title'       => __( 'Pineapple Pizza', 'tribe' ),
			'description' => __( 'A succesful run of the Tribe Libs block generator', 'tribe' ), // TODO: describe the block
			'icon'        => '<svg width="28" height="19" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M8.254 5.968H19.85v1.054H8.254V5.968zm1.462 1.57h8.568v1.053H9.716V7.538z" fill="#000"/><path d="M2.82 2.437v14.126H25.2V2.437H2.82zm21.482 13.176H3.866V3.366h20.436v12.247z" fill="#000"/><path d="M10.092 16.15H2.884l10.614-3.242 1.693 1.012 5.433-3.243 4.409 2.623v2.85h-14.94z" fill="#000"/></svg>', // TODO: set SVG icon
			'keywords'    => [ __( 'placeholder', 'tribe' ) ], // TODO: select appropriate keywords
			'category'    => 'common', // core categories: common, formatting, layout, widgets, embed
			'supports'    => [
				'align'  => false,
				'anchor' => true,
				'html'   => false,
			],
		] ) );
	}

	/**
	 * Register Fields for block
	 */
	public function add_fields() {
		//==========================================
		// Content Fields
		//==========================================
		$this->add_section( new Field_Section( self::SECTION_CONTENT, __( 'Content', 'tribe' ), 'accordion' ) )
			->add_field( new Field( self::NAME . '_' . self::TITLE, [
					'label' => __( 'Title', 'tribe' ),
					'name'  => self::TITLE,
					'type'  => 'text',
				] )
			)->add_field( new Field( self::NAME . '_' . self::DESCRIPTION, [
					'label'        => __( 'Description', 'tribe' ),
					'name'         => self::DESCRIPTION,
					'type'         => 'wysiwyg',
					'toolbar'      => 'basic',
					'media_upload' => 0,
				] )
			);

		//==========================================
		// Setting Fields
		//==========================================
		$this->add_section( new Field_Section( self::SECTION_SETTINGS, __( 'Settings', 'tribe' ), 'accordion' ) );
	}
}

[Dry Run] Block model file: /application/www/wp-content/plugins/core/src/Blocks/Types/Pineapple_Pizza/Pineapple_Pizza_Model.php
Block model contents:
<?php
declare( strict_types=1 );

namespace Tribe\Project\Blocks\Types\Pineapple_Pizza;

use Tribe\Project\Templates\Components\blocks\pineapple_pizza\Pineapple_Pizza_Controller;

/**
 * Class Pineapple_Pizza_Model
 *
 * Responsible for mapping values from the block to arguments
 * for the component
 */
class Pineapple_Pizza_Model extends \Tribe\Project\Blocks\Types\Base_Model {
	public function get_data(): array {
		return [
			Pineapple_Pizza_Controller::CLASSES     => $this->get_classes(),
			// TODO: map values from the block to the controller
		];
	}
}

[Dry Run] Block template file: /application/www/wp-content/themes/core/blocks/pineapplepizza/pineapplepizza.php
Block template contents:
<?php
declare( strict_types=1 );

/**
 * @var array $args Arguments passed to the template
 */
// phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UndefinedVariable
$model = new \Tribe\Project\Blocks\Types\Pineapple_Pizza\Pineapple_Pizza_Model( $args['block'] );
get_template_part( 'components/blocks/pineapple_pizza/pineapple_pizza', null, $model->get_data() );

[Dry Run] Template file: /application/www/wp-content/themes/core/components/blocks/pineapple_pizza/pineapple_pizza.php
Template contents:
<?php
declare( strict_types=1 );

/**
 * @var array $args Arguments passed to the template
 */
// phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UndefinedVariable
$c = \Tribe\Project\Templates\Components\blocks\pineapple_pizza\Pineapple_Pizza_Controller::factory( $args );
?>
<div
	<?php echo $c->get_classes(); ?>
	<?php echo $c->get_attributes(); ?>
>

</div>

[Dry Run] Controller file: /application/www/wp-content/themes/core/components/blocks/pineapple_pizza/Pineapple_Pizza_Controller.php
Controller contents:
<?php
declare( strict_types=1 );

namespace Tribe\Project\Templates\Components\blocks\pineapple_pizza;

use Tribe\Libs\Utils\Markup_Utils;
use Tribe\Project\Templates\Components\Abstract_Controller;

class Pineapple_Pizza_Controller extends Abstract_Controller {

	public const CLASSES = 'classes';
	public const ATTRS   = 'attrs';

	/**
	 * @var string[]
	 */
	private array $classes;
	/**
	 * @var string[]
	 */
	private array $attrs;

	public function __construct( array $args = [] ) {
		$args = $this->parse_args( $args );

		$this->classes = (array) $args[ self::CLASSES ];
		$this->attrs   = (array) $args[ self::ATTRS ];
	}

	protected function defaults(): array {
		return [
			self::CLASSES => [],
			self::ATTRS   => [],
		];
	}

	protected function required(): array {
		return [];
	}

	public function get_classes(): string {
		return Markup_Utils::class_attribute( $this->classes );
	}

	public function get_attributes(): string {
		return Markup_Utils::concat_attrs( $this->attrs );
	}
}

[Dry Run] CSS index file: /application/www/wp-content/themes/core/components/blocks/pineapple_pizza/index.pcss
CSS index contents:
/* -----------------------------------------------------------------------------
 *
 * Component: Pineapple Pizza
 *
 * This file is just a clearing-house, see the css directory
 * and edit the source files found there.
 *
 * ----------------------------------------------------------------------------- */

@import "css/pineapple_pizza.pcss";

[Dry Run] CSS source file: /application/www/wp-content/themes/core/components/blocks/pineapple_pizza/css/pineapple_pizza.pcss
CSS source contents:
/* -----------------------------------------------------------------------------
 *
 * Component: Pineapple Pizza
 *
 * ----------------------------------------------------------------------------- */



[Dry Run] JS index file: /application/www/wp-content/themes/core/components/blocks/pineapple_pizza/index.js
JS index contents:
/* -----------------------------------------------------------------------------
 *
 * Component: Pineapple Pizza
 *
 * This file is just a clearing-house, see the css directory
 * and edit the source files found there.
 *
 * ----------------------------------------------------------------------------- */

import pineapple_pizza from './js/pineapple_pizza';

const init = () => {
	pineapple_pizza();
};

export default init;

[Dry Run] JS source file: /application/www/wp-content/themes/core/components/blocks/pineapple_pizza/js/pineapple_pizza.js
JS source contents:
/* -----------------------------------------------------------------------------
 *
 * Component: Pineapple Pizza
 *
 * ----------------------------------------------------------------------------- */
/**
 * @module
 * @description TODO: add a description to this file
 */

import * as tools from 'utils/tools';

/**
 * @function init
 * @description Kick off this module's functions
 */

const init = () => {
	console.info( 'Square One FE: Initialized Pineapple Pizza component scripts.' );
};

export default init;

Success: Way to go! pineapple_pizza component has been created
[Dry Run] Skipping registration of block type in Blocks_Definer.php
Success: Way to go! pineapplepizza block has been created
```

**Review by msawicki — APPROVED** [review]
This is amazing Jonathan, thanks for putting this together.

File: src/Generators/templates/block/config.php.tmpl [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	public const SECTION_SETTINGS = 's-settings';
+
+	public const TITLE       = 'title';
+	public const DESCRIPTION = 'description';
+
+	public function add_block() {
+		$this->set_block( new Block( self::NAME, [
+			'title'       => __( '%3$s', 'tribe' ),
+			'description' => __( 'A succesful run of the Tribe Libs block generator', 'tribe' ), // TODO: describe the block
+			'icon'        => '<svg width="28" height="19" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M8.254 5.968H19.85v1.054H8.254V5.968zm1.462 1.57h8.568v1.053H9.716V7.538z" fill="#000"/><path d="M2.82 2.437v14.126H25.2V2.437H2.82zm21.482 13.176H3.866V3.366h20.436v12.247z" fill="#000"/><path d="M10.092 16.15H2.884l10.614-3.242 1.693 1.012 5.433-3.243 4.409 2.623v2.85h-14.94z" fill="#000"/></svg>', // TODO: set SVG icon
```
[the remark this answers is not in the document]
Comment (jbrinley) [reply src/Generators/templates/block/config.php.tmpl:24]: image

[the remark this answers is not in the document]
Comment (jbrinley) [reply src/Generators/templates/block/config.php.tmpl:24]: On reflection, maybe this is a better default:
image

**Pull request merged**